### ISSUE TITLE
Remove dns-prefetch and assets loaded from s.w.org

### ISSUE THREAD

**Issue by stefanos82** [ClassicPress/ClassicPress#140]
I have found two places that mention `dns-prefetch` that points at `s.w.org` which redirects to WordPress's website.

Also, it can be found in ClassicPress's website as `<link rel='dns-prefetch' href='//s.w.org' />`.

Is there a particular reason we need it?

Please advise.

Thank you.

**Comment by nylen** [comment]
Good catch!

```
$ ack s.w.org
tests/phpunit/tests/formatting/Emoji.php:9:	private $png_cdn = 'https://s.w.org/images/core/emoji/11/72x72/';
tests/phpunit/tests/formatting/Emoji.php:10:	private $svn_cdn = 'https://s.w.org/images/core/emoji/11/svg/';
tests/phpunit/tests/http/base.php:16:	var $fileStreamUrl = 'http://s.w.org/screenshots/3.9/dashboard.png';
tests/phpunit/tests/http/base.php:35:		if ( 0 === strpos( $response->get_error_message(), 'stream_socket_client(): unable to connect to tcp://s.w.org:80' ) ) {
tests/phpunit/tests/general/resourceHints.php:34:		$expected = "<link rel='dns-prefetch' href='//s.w.org' />\n";
tests/phpunit/tests/general/resourceHints.php:42:		$expected = "<link rel='dns-prefetch' href='//s.w.org' />\n" .
tests/phpunit/tests/general/resourceHints.php:73:		$expected = "<link rel='dns-prefetch' href='//s.w.org' />\n" .
tests/phpunit/tests/general/resourceHints.php:101:		$expected = "<link rel='dns-prefetch' href='//s.w.org' />\n" .
tests/phpunit/tests/general/resourceHints.php:127:		$expected = "<link rel='dns-prefetch' href='//s.w.org' />\n" .
tests/phpunit/tests/general/resourceHints.php:149:					"<link rel='dns-prefetch' href='//s.w.org' />\n";
tests/phpunit/tests/general/resourceHints.php:168:					"<link rel='dns-prefetch' href='//s.w.org' />\n";
tests/phpunit/tests/general/resourceHints.php:185:		$expected = "<link rel='dns-prefetch' href='//s.w.org' />\n";
tests/phpunit/tests/general/resourceHints.php:202:		$expected = "<link rel='dns-prefetch' href='//s.w.org' />\n";
tests/phpunit/tests/general/resourceHints.php:215:		$expected = "<link rel='dns-prefetch' href='//s.w.org' />\n";
tests/phpunit/tests/general/resourceHints.php:250:		$expected = "<link rel='dns-prefetch' href='//s.w.org' />\n" .
tests/qunit/wp-admin/js/widgets/test-media-image-widget.js:13:		testImageUrl = 'http://s.w.org/style/images/wp-header-logo.png';
tests/qunit/wp-admin/js/widgets/test-media-image-widget.js:94:		imageWidgetControlInstance.model.set({ error: false, url: 'http://s.w.org/style/images/wp-header-logo.png' });
tests/qunit/wp-admin/js/widgets/test-media-image-widget.js:98:			equal( imageWidgetControlInstance.$el.find( 'img[src="http://s.w.org/style/images/wp-header-logo.png"]' ).length, 1, 'One image should be rendered' );
src/wp-includes/formatting.php:5036:		'baseUrl' => apply_filters( 'emoji_url', 'https://s.w.org/images/core/emoji/11/72x72/' ),
src/wp-includes/formatting.php:5054:		'svgUrl' => apply_filters( 'emoji_svg_url', 'https://s.w.org/images/core/emoji/11/svg/' ),
src/wp-includes/formatting.php:5176:	$cdn_url = apply_filters( 'emoji_url', 'https://s.w.org/images/core/emoji/11/72x72/' );
src/wp-includes/general-template.php:2888:	$hints['dns-prefetch'][] = apply_filters( 'emoji_svg_url', 'https://s.w.org/images/core/emoji/11/svg/' );
src/wp-admin/about.php:219:				<source media="(max-width: 500px)" srcset="<?php echo 'https://s.w.org/images/core/4.9/banner-mobile.svg'; ?>">
src/wp-admin/about.php:220:				<img src="https://s.w.org/images/core/4.9/banner.svg" alt="">
src/wp-admin/about.php:232:						<img src="https://s.w.org/images/core/4.9/draft-and-schedule.svg" alt="">
src/wp-admin/about.php:239:						<img src="https://s.w.org/images/core/4.9/design-preview-links.svg" alt="">
src/wp-admin/about.php:246:						<img src="https://s.w.org/images/core/4.9/locking.svg" alt="">
src/wp-admin/about.php:253:						<img src="https://s.w.org/images/core/4.9/prompt.svg" alt="">
src/wp-admin/about.php:269:						<img src="https://s.w.org/images/core/4.9/syntax-highlighting.svg" alt="">
src/wp-admin/about.php:276:						<img src="https://s.w.org/images/core/4.9/sandbox.svg" alt="">
src/wp-admin/about.php:283:						<img src="https://s.w.org/images/core/4.9/warning.svg" alt="">
src/wp-admin/about.php:299:						<img src="https://s.w.org/images/core/4.9/gallery-widget.svg" alt="">
src/wp-admin/about.php:306:						<img src="https://s.w.org/images/core/4.9/media-button.svg" alt="">
src/wp-admin/about.php:322:						<img src="https://s.w.org/images/core/4.9/theme-switching.svg" alt="">
src/wp-admin/about.php:329:						<img src="https://s.w.org/images/core/4.9/menu-flow.svg" alt="">
```

So, emoji assets are served from this host, and so are a few things from the wp-admin About page (which we'll remove, see #4).

Probably the best solution for us for v1 is to bundle the emoji assets with ClassicPress and serve them directly from the install, but I'm open to ideas and contributions here.

See also: https://petitions.classicpress.net/posts/7/remove-emojis-gdpr-friendly

**Comment by stefanos82** [comment]
@nylen I was about to suggest the same thing; just bundle them for now and afterwards we can serve them as part of a service that will handle ClassicPress's static content as a whole.

**Comment by stefanos82** [comment]
@nylen I have checked with silver searcher `ag` and have found files that contain WordPress links. I'm going _to list just a few_ of them here and of course I'm pretty sure there are some that could be considered as false-positives, such as links inside comments and so forth; nevertheless, feel free to check them yourself:

- tools/i18n/makepot.php
- Gruntfile.js
- tools/i18n/t/data/not-gettexted-0-result.php
- src/wp-includes/customize/class-wp-customize-nav-menu-locations-control.php
- src/wp-includes/customize/class-wp-customize-themes-section.php
- tools/i18n/t/data/not-gettexted-0.po
- src/wp-includes/class-oembed.php
- src/wp-includes/query.php
- src/wp-includes/shortcodes.php
- src/wp-includes/comment-template.php
- src/wp-includes/general-template.php
- src/wp-includes/theme.php
- src/wp-includes/nav-menu.php
- src/wp-includes/default-constants.php
- src/wp-includes/wp-db.php
- src/wp-includes/formatting.php
- src/wp-includes/deprecated.php
- src/wp-includes/class-wp-theme.php
- src/wp-includes/ms-load.php
- src/wp-includes/class-wp-editor.php
- src/wp-includes/js/mce-view.js

The list could go on...

**Comment by nylen** [comment]
Let's focus on `s.w.org` for the purpose of this issue, that is nice and specific.  If you'd like to create other issues for other specific types of links we should look at, then please do.

**Comment by stefanos82** [comment]
Not a bad idea actually. I will do that right now.

**Comment by stefanos82** [comment]
My suggestion would be to create a new parent folder named `assets` inside root directory with sub-folders `images/core/` and place them there.

As soon as we set up a service URL, it should point at `https://foo-something.org/images/core/...`.

The reason I have suggested `assets` is so we can move other types of static files as well, such as CSS, small videos if we like...there are many benefits to get out of it.

**Comment by nylen** [comment]
> **Where will these images be stored within the local file structure?**

> 1. wp-admin/about.php

These were specific to the WP About page.  These references have all been removed from the ClassicPress code so there is nothing to do here.

> 2. https://s.w.org/images/core/emoji/11/

Let's put these under `wp-includes/images/emoji/`.  I am also curious what the `11` means in those URLs.

> 3. Will we repoint or remove the DNS prefetch?

After `s.w.org` references are removed, we can remove the DNS prefetch too.  There's no point in adding a DNS prefetch to the same site you're currently on, as we already know that the browser knows how to reach it :wink: 

> My suggestion would be to create a new parent folder named `assets` inside root directory

There are already places for all of these types of assets in the current code structure (`images`, `css`, `js` folders under `wp-includes`).

I'm definitely not saying the current structure isn't ideal, but let's think about a more complete reorganization for a future major version (v2 or beyond), and until then, avoid making changes unless absolutely necessary.

**Comment by senlin** [comment]
I have time to do this, thing is though that the emoji reference URLs are all blocked?

**Comment by nylen** [comment]
What do you mean by blocked?

**Comment by senlin** [comment]
The 4 s.w.org URLs are blocked as in:

<img width="698" alt="blocked-sworg" src="https://user-images.githubusercontent.com/650066/47968262-5eb16880-e05f-11e8-83aa-061b57f243e7.png">

For the following URLs:
* https://s.w.org/images/core/emoji/11/72x72/ (baseUrl)
* https://s.w.org/images/core/emoji/11/svg/ (svgUrl)
* https://s.w.org/images/core/emoji/11/72x72/ ($cdn_url)
* https://s.w.org/images/core/emoji/11/svg/ (DNS prefetch)

**Comment by nylen** [comment]
Ah. More specifically: they don't have directory indexes enabled there.

There are files included from s.w.org which are valid, though. So part of this task is to find a full list of those files and mirror them.

**Comment by senlin** [comment]
> So part of this task is to find a full list of those files and mirror them.

Guess I will put the Help wanted label back then.

**Comment by senlin** [comment]
@nylen would it be safe to already remove the references from the tests?

**Comment by nylen** [comment]
Thanks @kraftbj!

So maybe `src/wp-includes/images/twemoji/11/72x72/` and `src/wp-includes/images/twemoji/11/svg/` is the way to go here.

**Comment by senlin** [comment]
Question: 

After this is merged, this

```
	private $png_cdn = 'https://s.w.org/images/core/emoji/11/72x72/';
	private $svn_cdn = 'https://s.w.org/images/core/emoji/11/svg/';
```

will have to become something like

`dirname( dirname( dirname( __FILE__ ) ) ) . '/src/wp-includes/core/emoji/11/..`

?

**Comment by senlin** [comment]
By the way, this adds no less than 8MB to the install, is that really necessary? Do business sites need emojis in core?

**Comment by nylen** [comment]
We need to make sure that we point the new files at a valid URL.

See also discussion on #212 - I'm thinking we will want to serve the SVG emojis only, using a php file that returns the right image based on a query string argument.

This will be a pretty large PHP file, but at least it will compress down to a really small size in the build.

**Comment by nylen** [comment]
> is that really necessary? Do business sites need emojis in core?

I agree with you that sites should be able to disable this easily, and I think this part is a better fit for moving to a core plugin in v2.

We could also look at adding (an) option(s) to disable specific emoji-related functionality, but I'd like to get some other people's opinions on that first.

**Comment by senlin** [comment]
I already removed the PNG files from the PR, how do you want to proceed with this?

- close the PR, merge SVGs into 1 PHP file, write query and then do a new PR
- do the work above in/with the current PR
- accept the PR and do merging and writing of query at a later point (but before b1)
- something else

**Comment by nylen** [comment]
We can do the work above in the current PR.

About the first option: The existing PR is still a good starting point and it would mean more work to close it.

About the third option: It's not ideal to be merging parts of unfinished features that leave "loose ends" or bugs in the code. Ideally we'll merge the changes here all at once.

**Comment by senlin** [comment]
Ok great, thanks for explaining that :)

Now the next questions are:
1. how to merge the 2800+ SVG files into 1PHP file?
2. how to write a query that pulls the correct SVG out of that new PHP file?

re 1) I have been looking around on how to merge SVG files but cannot seem to find anything relevant
re 2) let's get 1 done first :)

**Comment by nylen** [comment]
Moved to beta3 milestone due to lack of time.

**Comment by nylen** [comment]
@invisnet has been working on this and he realized that there were some significant drawbacks with the approach I proposed above.

- Putting all of this SVG data (~5 MB uncompressed) inside a PHP file would probably not be very good for [PHP op caching systems](https://secure.php.net/manual/en/intro.opcache.php) as this would overrun the cache with a bunch of irrelevant data.
- We could avoid that, but this adds ~1 MB file to the final size of the zip file.

There is a much less complicated way to handle this issue for v1 - namely, [host the emoji files on ClassicPress servers](https://twemoji.classicpress.net/11/svg/e50a.svg).  Then it's just a matter of changing `s.w.org` to point at our server instead.

The code related to emojis will be moved to a plugin early in v2 - this is currently the top-rated [ClassicPress petition](https://petitions.classicpress.net/?view=most-wanted):

![2019-02-18t14 52 57-05 00](https://user-images.githubusercontent.com/227022/52973707-ef9d1d00-338c-11e9-8412-3b91f4c5e41a.png)

Since this is going away soon, let's not spend any more time re-working this code than we have to.

PR #212 has been updated with this approach instead, I'm finishing a couple of final steps with the tests and then merging it.
